### PR TITLE
[virtual_nfts][6] Refactor usage of MerkleMap<EntityNumPair, MerkleUniqueToken> to use an intermediate adaptor.

### PR DESCRIPTION
**Pull request by rjuang**
This PR creates a UniqueTokenMapAdapter which serves as a replacement for MerkleMap<EntityNumPair, MerkleUniqueToken>. The adapter class will be used as a drop-in to quickly switch between accessing the virtual vs non-virtualized NFT maps. This will enable guarding virtualizing NFTs behind a feature flag.

Related issue: https://github.com/hashgraph/hedera-services/issues/3089

Note for reviewers:
- Review changes in `UniqueTokenMapAdapter.java` and `UniqueTokenAdapter.java` first.
- All remaining changes correspond to replacing `MerkleMap<EntityNumPair, MerkleUniqueToken>` with `UniqueTokenMapAdapter`

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/ServicesState.java [hunk 1]
```diff
@@ -237,18 +242,24 @@ private void deserializedInit(
                     new MerkleScheduledTransactions(
                             ((MerkleMap<?, ?>) getChild(StateChildIndices.SCHEDULE_TXS)).size());
         }
-
-        internalInit(platform, new BootstrapProperties(), dualState, trigger, deserializedVersion);
+        final var bootstrapProps = new BootstrapProperties();
+        enabledVirtualNft =
+                bootstrapProps.getBooleanProperty(PropertyNames.TOKENS_NFTS_USE_VIRTUAL_MERKLE);
```
[the remark this answers is not in the document]
Comment (rjuang) [reply hedera-node/src/main/java/com/hedera/services/ServicesState.java:247]: yes. It's used in the follow-up PR.

https://github.com/hashgraph/hedera-services/pull/3843/files#diff-1ca5738f0fe42a25f91d641aea2104dd51bec20398ce6c2aa18afde3d5a590d9

under ServicesState.shouldMigrateNfts()

File: hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java [hunk 2]
```diff
@@ -338,8 +338,7 @@ private Key grpcKeyReprOf(final byte[] publicKey) {
     public Optional<TokenNftInfo> infoForNft(final NftID target) {
         final var currentNfts = uniqueTokens();
         final var tokenId = EntityNum.fromTokenId(target.getTokenID());
-        final var targetKey =
-                EntityNumPair.fromLongs(tokenId.longValue(), target.getSerialNumber());
+        final var targetKey = NftId.fromGrpc(target);
```
[the remark this answers is not in the document]
Comment (rjuang) [reply hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java:341]: Oh. I had not thought about that. I can consider using EntityNumPair then.

[the remark this answers is not in the document]
Comment (rjuang) [reply hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java]: better yet, I can avoid using fromGrpc and just use another factory constructor. Looks like NftId is just a plain record.

**Review by Neeharika-Sompalli — DISMISSED** [review]
LGTM ! Will let @tinker-michaelj  review as well.

File: hedera-node/src/main/java/com/hedera/services/ServicesState.java [hunk 3]
```diff
@@ -464,8 +475,13 @@ public RecordsRunningHashLeaf runningHashLeaf() {
         return getChild(StateChildIndices.RECORD_STREAM_RUNNING_HASH);
     }
 
-    public MerkleMap<EntityNumPair, MerkleUniqueToken> uniqueTokens() {
-        return getChild(StateChildIndices.UNIQUE_TOKENS);
+    public UniqueTokenMapAdapter uniqueTokens() {
+        final var tokensMap = getChild(StateChildIndices.UNIQUE_TOKENS);
+        return tokensMap.getClass() == MerkleMap.class
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/ServicesState.java:480]: Is there a reason to not just do `tokensMap instanceof MerkleMap`?

File: hedera-node/src/main/java/com/hedera/services/state/expiry/UniqueTokensListMutation.java [hunk 4]
```diff
@@ -17,75 +17,73 @@
...
-    public void markAsTail(final MerkleUniqueToken node) {
+    public void markAsTail(final UniqueTokenAdapter node) {
         node.setNext(MISSING_NFT_NUM_PAIR);
     }
 
     @Override
-    public void updatePrev(final MerkleUniqueToken node, final EntityNumPair prev) {
-        node.setPrev(prev.asNftNumPair());
+    public void updatePrev(final UniqueTokenAdapter node, final NftId prev) {
+        node.setPrev(prev.asEntityNumPair().asNftNumPair());
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/state/expiry/UniqueTokensListMutation.java:68]: Is it worth having a direct `prev.asNftNumPair()` view?

File: hedera-node/src/main/java/com/hedera/services/context/ImmutableStateChildren.java [hunk 5]
```diff
@@ -51,7 +52,8 @@ public class ImmutableStateChildren implements StateChildren {
     private final WeakReference<MerkleMap<EntityNum, MerkleAccount>> accounts;
     private final WeakReference<MerkleMap<EntityNum, MerkleTopic>> topics;
     private final WeakReference<MerkleMap<EntityNum, MerkleToken>> tokens;
-    private final WeakReference<MerkleMap<EntityNumPair, MerkleUniqueToken>> uniqueTokens;
+    // UniqueTokenMapAdapter is constructed on demand, so a strong reference needs to be held.
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/context/ImmutableStateChildren.java:55]: I'm having trouble grasping this comment. 😝 

What happens if this is a `WeakReference`?

**Review by tinker-michaelj — APPROVED** [review]
LGTM!

File: hedera-node/src/main/java/com/hedera/services/ServicesState.java (hunk 3 above)
Comment (rjuang) [reply hedera-node/src/main/java/com/hedera/services/ServicesState.java:480]: `instanceof` is a very expensive operation (at least this was true in Java 8) where as the equality check for .getClass() is not as expensive. That was the reason for avoiding instanceof

File: hedera-node/src/main/java/com/hedera/services/state/expiry/UniqueTokensListMutation.java (hunk 4 above)
Comment (rjuang) [reply hedera-node/src/main/java/com/hedera/services/state/expiry/UniqueTokensListMutation.java:68]: yea.. that would be a good idea. Let me take a look at modifying this.

File: hedera-node/src/main/java/com/hedera/services/context/ImmutableStateChildren.java (hunk 5 above)
Comment (rjuang) [reply hedera-node/src/main/java/com/hedera/services/context/ImmutableStateChildren.java:55]: If it's a weak reference, then there would be no one holding on to the reference constructed by:
`UniqueTokenAdapter.wrap(token)` that gets provided to the instance when set.

Comment (tinker-michaelj) [reply hedera-node/src/main/java/com/hedera/services/context/ImmutableStateChildren.java:55]: Does this mean it's actually safe to drop all the `WeakReference`'s here?

**Review by tinker-michaelj — APPROVED** [review]
LGTM

**Review by Neeharika-Sompalli — APPROVED** [review]
LGTM !

**Pull request merged**